kobotoolbox/kpi
# Log project name and settings changes

## Checklist

1. [x] If you've added code that should be tested, add tests
2. [ ] If you've changed APIs, update (or create!) the documentation
3. [x] Ensure the tests pass
4. [x] Run `./python-format.sh` to make sure that your code lints and that you've followed [our coding style](https://github.com/kobotoolbox/kpi/blob/main/CONTRIBUTING.md)
5. [x] Write a title and, if necessary, a description of your work suitable for publishing in our [release notes](https://community.kobotoolbox.org/tag/release-notes)
6. [ ] Mention any related issues in this repository (as #ISSUE) and in other repositories (as kobotoolbox/other#ISSUE)
7. [ ] Open an issue in the [docs](https://github.com/kobotoolbox/docs/issues/new) if there are UI/UX changes
8. [x] Create a testing plan for the reviewer and add it to the Testing section

## Description

Create project history logs for changes to project name and settings.

## Notes

Refactor the asset model to be more explicit about which settings fields are standardized for saving (makes testing easier)
1. Add `settings` and `latest_version.uid` to the fields tracked on the AssetViewSet requests
2. Add a method to ProjectHistoryLog that handles an `update` request by determining which fields were changed and delegating to an appropriate handler. 
3. Add handlers  to deal with changes to name and settings
   a. For settings, if the setting is a list, include values added and values removed, otherwise, just store the old and new values


## Testing
1. Create a new project
2. Go to Settings > General
3. Update the name and save
4. Go to `/api/v2/audit-logs/?q=metadata__asset_uid=<project uid> AND log_type='project-history'&format=json`
5. You should see a project history log that looks like:
```
        {
            "app_label": "kpi",
            "model_name": "asset",
            "user": "http://kf.kobo.local:8080/api/v2/users/{you}/",
            "user_uid": "{your uid}",
            "username": "{your username}",
            "action": "update-name",
            "metadata": {
                "source": "{browser} ({OS})",
                "asset_uid": "{new project uid}",
                "ip_address": "172.18.0.1",
                "log_subtype": "project",
                "latest_version_uid": "{version uid}",
                "name": {
                      "old": <old_name>,
                      "new": <new_name>,
                }
            },
            "date_created": "{whenever}",
            "log_type": "project-history"
        }

```
6. Update the description to 'new description' and countries to Albania and Algeria and save
7. Reload the endpoint. You should see a new project history log with:
    a. action = 'update-settings'
    b. metadata.settings = 
        ```
        {
              'description': { 'old': '', 'new': 'new description'},
              'country': {'added': [{'label':'Albania', 'value':'ALB'}, {'label': 'Algeria', 'value': 'ALG'},
                                  'removed': []},
              'country_codes': {'added': ['ALB','ALG'], 'removed': []}
        }
        ```
8. Update the name (don't save yet)
9. Update the sector to 'Environment' and save
10. Reload the endpoint. You should see 2 new project history logs, one for the name change and one for the settings change
11. Go to edit the form
12. Change the name of the project from the form editing screen
13. Reload the endpoint and make sure there's a new log for the name change
14. To test deleting/resetting the usual settings: run `curl -X PATCH -H 'Authorization: Token <your token>'  -H 'Content-Type: application/json' -H 'Accept: application/json' http://kf.kobo.local/api/v2/assets/<asset uid> --data '{"settings": {}}'`
15. Reload the endpoint. You should see a new project history log with `metadata.settings` =
     ```
     {
          'country': {
                'added': [],
                'removed': [{'label':'Albania', 'value':'ALB'}, {'label':'Algeria', 'value': 'ALG'}], 
          }
          'country_codes': {
                 'added': [],
                 'removed': ['ALB', 'ALG']
          }
          'description': {
                  'old': 'new description',
                  'new': '',
           }
          'sector': {
                  'old': {'label': 'environment', 'value':'environment'},
                  'new': {}
          }
     }
     ```
16. To test new settings: run `curl -X PATCH -H 'Authorization: Token <your token>'  -H 'Content-Type: application/json' -H 'Accept: application/json' http://kf.kobo.local/api/v2/assets/<asset uid> --data '{"settings": {"new_setting": "new_value"}}'`
17. Reload the endpoint. You should see a new project history log with `metadata.settings` =
     ```
     {
           'new_setting': {
                 'old': null,
                 'new': 'new_value',
            }
     }
     ```
18. To test deleting an arbitrary setting: run `curl -X PATCH -H 'Authorization: Token <your token>'  -H 'Content-Type: application/json' -H 'Accept: application/json' http://kf.kobo.local/api/v2/assets/<asset uid> --data '{"settings": {}}'`
19. Reload the endpoint. You should see a new project history log with `metadata.settings` =
     ```
     {
           'new_setting': {
                 'old': 'new_value',
                 'new': null,
            }
     }
     ```
